inject_test.go

```diff
@@ -0,0 +1,185 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/dig"
+)
+
+func TestInject(t *testing.T) {
+	type type1 struct{}
+	type type2 struct{}
+	type type3 struct{}
+
+	t.Run("Failures", func(t *testing.T) {
+		tests := []interface{}{
+			3,
+			func() {},
+			struct{}{},
+			struct{ Foo *bytes.Buffer }{},
+		}
+
+		for _, tt := range tests {
+			t.Run(fmt.Sprintf("%T", tt), func(t *testing.T) {
+				app := New()
+				app.Provide(&bytes.Buffer{})
+
+				err := app.Start(context.Background(), Inject(&tt))
+				require.Error(t, err, "expected failure")
+				require.Contains(t, err.Error(), "Inject expected a pointer to a struct")
+			})
+		}
+	})
+
+	t.Run("Empty", func(t *testing.T) {
+		new1 := func() *type1 { panic("new1 must not be called") }
+		new2 := func() *type2 { panic("new2 must not be called") }
+
+		app := New()
+		app.Provide(new1, new2)
+
+		var out struct{}
+		require.NoError(t, app.Start(context.Background(), Inject(&out)), "start failed")
+	})
+
+	t.Run("StructIsInjected", func(t *testing.T) {
+		var gave1 *type1
+		new1 := func() *type1 {
+			gave1 = &type1{}
+			return gave1
+		}
+
+		var gave2 *type2
+		new2 := func() *type2 {
+			gave2 = &type2{}
+			return gave2
+		}
+
+		app := New()
+		app.Provide(new1, new2)
+
+		var out struct {
+			T1 *type1
+			T2 *type2
+		}
+		require.NoError(t, app.Start(context.Background(), Inject(&out)),
+			"failed to start")
+
+		assert.NotNil(t, out.T1, "T1 must not be nil")
+		assert.NotNil(t, out.T2, "T2 must not be nil")
+		assert.True(t, gave1 == out.T1, "T1 must match")
+		assert.True(t, gave2 == out.T2, "T2 must match")
+	})
+
+	t.Run("SkipsUnexported", func(t *testing.T) {
+		var gave1 *type1
+		new1 := func() *type1 {
+			gave1 = &type1{}
+			return gave1
+		}
+
+		new2 := func() *type2 { panic("new2 must not be called") }
+
+		var gave3 *type3
+		new3 := func() *type3 {
+			gave3 = &type3{}
+			return gave3
+		}
+
+		app := New()
+		app.Provide(new1, new2, new3)
+
+		var out struct {
+			T1 *type1
+			t2 *type2
+			T3 *type3
+		}
+		require.NoError(t, app.Start(context.Background(), Inject(&out)),
+			"failed to start")
+
+		assert.NotNil(t, out.T1, "T1 must not be nil")
+		assert.Nil(t, out.t2, "t2 must be nil")
+		assert.NotNil(t, out.T3, "T3 must not be nil")
+
+		assert.True(t, gave1 == out.T1, "T1 must match")
+		assert.True(t, gave3 == out.T3, "T3 must match")
+	})
+
+	t.Run("DoesNotZeroUnexported", func(t *testing.T) {
+		var gave1 *type1
+		new1 := func() *type1 {
+			gave1 = &type1{}
+			return gave1
+		}
+
+		new2 := func() *type2 { panic("new2 must not be called") }
+
+		app := New()
+		app.Provide(new1, new2)
+
+		var out struct {
+			T1 *type1
```

Comment:
It would be stomped upon. There's no opt-out mechanism for exported fields. I'll add a test for this.